File: sdk/python/examples/test_mcp_servers.py
#!/usr/bin/env python3
"""
A3S Code Python SDK - MCP Servers Integration Tests

Tests:
1. Global MCP servers loaded from config.hcl
2. Session-level dynamic MCP server injection
3. Both global + session MCP servers working together
4. LLM autonomously selecting and calling MCP tools
5. Disabled MCP server is skipped

Run with: python examples/test_mcp_servers.py
"""

import asyncio
import os
import tempfile
from pathlib import Path
from a3s_code import Agent


CONFIG_PATH = str(
    Path(__file__).parent.parent.parent.parent.parent.parent / ".a3s" / "config.hcl"
)


def truncate(text, max_len=300):
    if not text or len(text) <= max_len:
        return text
    return f"{text[:max_len]}... (truncated)"


# ── Test 1: Global MCP servers from config.hcl ──────────────────────────────

async def test_global_mcp_from_config():
    """
    Write a temp config that extends the real config with a global mcp_servers block,
    then verify the agent auto-loads it and the LLM can call the tool.
    """
    print("\n🌐 Test 1: Global MCP Servers from config.hcl")
    print("-" * 80)

    # Read the real config and append an mcp_servers block
    real_config = Path(CONFIG_PATH).read_text()
    config_with_mcp = real_config + """
mcp_servers {
    name      = "everything"
    transport = "stdio"
    command   = "npx"
    args      = ["-y", "@modelcontextprotocol/server-everything", "stdio"]
    enabled   = true
}
"""

    with tempfile.NamedTemporaryFile(mode="w", suffix=".hcl", delete=False) as f:
        f.write(config_with_mcp)
        tmp_path = f.name

    try:
        agent = Agent.create(tmp_path)
        session = agent.session(".")

        print("Asking agent to use the MCP 'echo' tool autonomously...")
        result = session.send(
            "Use the mcp__everything__echo tool to echo the message 'Hello from global MCP!'"
        )
        print(f"✓ Response: {truncate(result.text)}")

        assert "Hello from global MCP" in result.text or "echo" in result.text.lower(), \
            f"Expected echo result, got: {result.text}"

        print("\n✅ Test 1 passed: Global MCP servers auto-loaded from config")
    finally:
        os.unlink(tmp_path)


# ── Test 2: Session-level dynamic MCP injection ──────────────────────────────

async def test_session_level_mcp_injection():
    """
    Verifies that add_mcp_server() on a live session registers tools
    immediately and the LLM can call them in the same session.
    """
    print("\n🔌 Test 2: Session-level Dynamic MCP Injection")
    print("-" * 80)

    agent = Agent.create(CONFIG_PATH)
    session = agent.session(".")

    print("Injecting MCP server dynamically into live session...")
    tool_count = session.add_mcp_server(
        name="everything",
        command="npx",
        args=["-y", "@modelcontextprotocol/server-everything", "stdio"],
    )
    print(f"✓ Registered {tool_count} tools from 'everything' server")
    assert tool_count > 0, "Expected at least 1 tool from server-everything"

    print("\nAsking agent to use the dynamically injected MCP tool...")
    result = session.send(
        "Use the mcp__everything__add tool to add the numbers 17 and 25"
    )
    print(f"✓ Response: {truncate(result.text)}")

    assert "42" in result.text, f"Expected 42 (17+25), got: {result.text}"

    print("\n✅ Test 2 passed: Session-level MCP injection works")


# ── Test 3: Global + Session MCP working together ────────────────────────────

async def test_global_and_session_mcp_together():
    """
    Verifies that global MCP servers (from config) and session-level
    MCP servers (from add_mcp_server) are both available simultaneously.
    """
    print("\n🔗 Test 3: Global + Session MCP Together")
    print("-" * 80)

    real_config = Path(CONFIG_PATH).read_text()
    config_with_global = real_config + """
mcp_servers {
    name      = "global-everything"
    transport = "stdio"
    command   = "npx"
    args      = ["-y", "@modelcontextprotocol/server-everything", "stdio"]
    enabled   = true
}
"""

    with tempfile.NamedTemporaryFile(mode="w", suffix=".hcl", delete=False) as f:
        f.write(config_with_global)
        tmp_path = f.name

    try:
        agent = Agent.create(tmp_path)
        session = agent.session(".")

        print("Injecting additional session-level MCP server...")
        tool_count = session.add_mcp_server(
            name="session-everything",
            command="npx",
            args=["-y", "@modelcontextprotocol/server-everything", "stdio"],
        )
        print(f"✓ Session server registered {tool_count} tools")

        print("\nAsking agent to use tools from BOTH servers...")
        result = session.send(
            "First use mcp__global-everything__echo to echo 'from global', "
            "then use mcp__session-everything__echo to echo 'from session'. "
            "Report both results."
        )
        print(f"✓ Response: {truncate(result.text)}")

        lower = result.text.lower()
        assert "global" in lower or "session" in lower, \
            f"Expected both MCP results, got: {result.text}"

        print("\n✅ Test 3 passed: Global and session MCP servers work together")
    finally:
        os.unlink(tmp_path)


# ── Test 4: LLM autonomously selects MCP tool ────────────────────────────────

async def test_llm_autonomous_mcp_selection():
    """
    Verifies that the LLM can autonomously decide to use an MCP tool
    without being explicitly told which tool to call.
    """
    print("\n🤖 Test 4: LLM Autonomous MCP Tool Selection")
    print("-" * 80)

    real_config = Path(CONFIG_PATH).read_text()
    config_with_mcp = real_config + """
mcp_servers {
    name      = "everything"
    transport = "stdio"
    command   = "npx"
    args      = ["-y", "@modelcontextprotocol/server-everything", "stdio"]
    enabled   = true
}
"""

    with tempfile.NamedTemporaryFile(mode="w", suffix=".hcl", delete=False) as f:
        f.write(config_with_mcp)
        tmp_path = f.name

    try:
        agent = Agent.create(tmp_path)
        session = agent.session(".")

        print("Asking agent to add numbers (without specifying which tool)...")
        result = session.send(
            "What is 123 plus 456? Use the available MCP tools to compute this."
        )
        print(f"✓ Response: {truncate(result.text)}")

        assert "579" in result.text, f"Expected 579 (123+456), got: {result.text}"

        print("\n✅ Test 4 passed: LLM autonomously selected and called MCP tool")
    finally:
        os.unlink(tmp_path)


# ── Test 5: Disabled MCP server is skipped ───────────────────────────────────

async def test_disabled_mcp_server_skipped():
    """
    Verifies that mcp_servers with enabled = false are not connected.
    """
    print("\n🚫 Test 5: Disabled MCP Server is Skipped")
    print("-" * 80)

    real_config = Path(CONFIG_PATH).read_text()
    config_with_disabled = real_config + """
mcp_servers {
    name      = "disabled-server"
    transport = "stdio"
    command   = "npx"
    args      = ["-y", "@modelcontextprotocol/server-everything", "stdio"]
    enabled   = false
}
"""

    with tempfile.NamedTemporaryFile(mode="w", suffix=".hcl", delete=False) as f:
        f.write(config_with_disabled)
        tmp_path = f.name

    try:
        agent = Agent.create(tmp_path)
        session = agent.session(".")

        print("Asking agent about available MCP tools...")
        result = session.send(
            "List all available tools that start with 'mcp__'. "
            "If there are none, say 'no MCP tools available'."
        )
        print(f"✓ Response: {truncate(result.text)}")

        assert "mcp__disabled" not in result.text.lower(), \
            f"Disabled server should not register tools, got: {result.text}"

        print("\n✅ Test 5 passed: Disabled MCP server correctly skipped")
    finally:
        os.unlink(tmp_path)


# ── Main ──────────────────────────────────────────────────────────────────────

async def main():
    print("🚀 A3S Code Python SDK - MCP Servers Integration Tests\n")
    print("=" * 80)
    print(f"📄 Using config: {CONFIG_PATH}")
    print("=" * 80)

    if not Path(CONFIG_PATH).exists():
        print(f"❌ Config not found: {CONFIG_PATH}")
        return

    tests = [
        test_global_mcp_from_config,
        test_session_level_mcp_injection,
        test_global_and_session_mcp_together,
        test_llm_autonomous_mcp_selection,
        test_disabled_mcp_server_skipped,
    ]

    passed = 0
    failed = 0
    for test in tests:
        try:
            await test()
            passed += 1
        except Exception as e:
            print(f"\n❌ FAILED: {e}")
            failed += 1

    print("\n\n" + "=" * 80)
    if failed == 0:
        print(f"✅ All {passed} MCP integration tests completed successfully!")
    else:
        print(f"⚠️  {passed} passed, {failed} failed")
    print("=" * 80)


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/node/examples/test_mcp_servers.js
#!/usr/bin/env node
/**
 * A3S Code Node.js SDK - Integration Tests
 *
 * Tests all recently added/fixed features against the real kimi endpoint:
 *   1. task tool (delegate to general subagent, wait for result)
 *   2. parallel_task (fan-out to multiple subagents concurrently)
 *   3. toolNames() initial state on a fresh session
 *   4. mcpStatus error field populated on failed connect
 *   5. MCP injection: add → status → LLM use → toolNames → remove
 *   6. refreshMcpTools (smoke test)
 *
 * Run with: node examples/test_mcp_servers.js
 * Requires: ~/.a3s/config.hcl or repo .a3s/config.hcl
 */

const { Agent } = require("../index.js");
const crypto = require("crypto");
const os = require("os");
const path = require("path");
const fs = require("fs");

// Bundled minimal MCP echo server — no external dependencies required
const ECHO_SERVER = path.join(__dirname, "mcp_echo_server.js");

function resolveConfig() {
  const homeConfig = path.join(os.homedir(), ".a3s", "config.hcl");
  if (fs.existsSync(homeConfig)) return homeConfig;
  let dir = __dirname;
  for (let i = 0; i < 10; i++) {
    const candidate = path.join(dir, ".a3s", "config.hcl");
    if (fs.existsSync(candidate)) return candidate;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  throw new Error("Config not found at .a3s/config.hcl or ~/.a3s/config.hcl");
}

function pass(label) {
  console.log(`  ✓  ${label}`);
}

// ── Test 1: task tool ─────────────────────────────────────────────────────────

async function testTaskTool(agent, tmpdir) {
  console.log("\n── Test 1: task tool (subagent delegation) ──");
  const session = agent.session(tmpdir, { permissive: true });
  const result = await session.send(
    "Use the task tool to delegate the following to the 'general' agent, " +
    "then return its exact reply verbatim: " +
    "'Reply with exactly the word: TASK_OK'"
  );
  if (!result.text.includes("TASK_OK")) {
    throw new Error(`task tool should return TASK_OK, got: ${result.text}`);
  }
  pass("task tool delegated and returned TASK_OK");
}

// ── Test 2: parallel_task ─────────────────────────────────────────────────────

async function testParallelTask(agent, tmpdir) {
  console.log("\n── Test 2: parallel_task (concurrent fan-out) ──");
  const session = agent.session(tmpdir, { permissive: true });
  const result = await session.send(
    "Use the parallel_task tool to run these three tasks concurrently " +
    "using the 'general' agent, then list their outputs:\n" +
    "1. Reply with exactly: PARALLEL_A\n" +
    "2. Reply with exactly: PARALLEL_B\n" +
    "3. Reply with exactly: PARALLEL_C"
  );
  for (const token of ["PARALLEL_A", "PARALLEL_B", "PARALLEL_C"]) {
    if (!result.text.includes(token)) {
      throw new Error(`expected ${token} in result, got: ${result.text}`);
    }
  }
  pass("parallel_task ran 3 subagents concurrently, all results returned");
}

// ── Test 3: toolNames initial state ──────────────────────────────────────────

async function testToolNames(agent, tmpdir) {
  console.log("\n── Test 3: toolNames() initial state ──");
  const session = agent.session(tmpdir, { permissive: true });

  const names = session.toolNames();
  if (names.length === 0) {
    throw new Error("expected built-in tools to be present");
  }
  pass(`toolNames() returns ${names.length} built-in tools`);

  const mcpNames = names.filter(n => n.startsWith("mcp__"));
  if (mcpNames.length > 0) {
    throw new Error(
      `expected no mcp__ tools on a fresh session (none configured globally), got: ${mcpNames}`
    );
  }
  pass("no mcp__ tools on fresh session (none configured globally)");
}

// ── Test 4: mcpStatus error field ─────────────────────────────────────────────

async function testMcpStatusError(agent, tmpdir) {
  console.log("\n── Test 4: mcpStatus error field on failed connect ──");
  const session = agent.session(tmpdir, { permissive: true });

  let err = null;
  try {
    await session.addMcpServer(
      "bad-server", "stdio", "nonexistent-mcp-binary-xyz", [], null, null, null
    );
  } catch (e) {
    err = e;
  }
  if (!err) throw new Error("addMcpServer must throw for a nonexistent binary");

  // The config is registered before connection is attempted, so the server
  // must always appear in mcpStatus — even on failure.
  const status = await session.mcpStatus();
  const badEntry = status.find(s => s.name === "bad-server");
  if (!badEntry) {
    throw new Error(
      `bad-server must appear in mcpStatus after failed connect, ` +
      `got names: ${status.map(s => s.name)}`
    );
  }
  if (badEntry.connected) {
    throw new Error("bad-server should not be connected");
  }
  if (!badEntry.error) {
    throw new Error(`error field must be populated after failed connect, got: ${JSON.stringify(badEntry)}`);
  }
  pass(`mcpStatus.error captured: ${badEntry.error}`);
}

// ── Test 5: MCP injection ─────────────────────────────────────────────────────

async function testMcpInjection(agent, tmpdir) {
  console.log("\n── Test 5: MCP injection (add → status → LLM use → toolNames → remove) ──");
  const session = agent.session(tmpdir, { permissive: true });

  // Before add: no mcp__echo__ tools
  const toolsBefore = session.toolNames();
  const echoToolsBefore = toolsBefore.filter(n => n.startsWith("mcp__echo__"));
  if (echoToolsBefore.length > 0) {
    throw new Error(`expected no mcp__echo__ tools before addMcpServer, got: ${echoToolsBefore}`);
  }
  pass("no mcp__echo__ tools before addMcpServer");

  // Generate a random secret unknown to the LLM — it can only be obtained by
  // actually calling mcp__echo__get_secret, preventing the LLM from faking it.
  const secret = crypto.randomBytes(8).toString("hex");

  // Add the bundled echo server (uses the same Node.js binary — no external deps)
  const count = await session.addMcpServer(
    "echo", "stdio",
    process.execPath, [ECHO_SERVER, secret],
    null, null, null
  );
  if (count === 0) throw new Error("echo server must expose >= 1 tool");
  pass(`addMcpServer registered ${count} tools`);

  // Verify mcpStatus: connected=true, toolCount=N, error=null
  const mcpStat = await session.mcpStatus();
  const echoStat = mcpStat.find(s => s.name === "echo");
  if (!echoStat) {
    throw new Error(`echo must appear in mcpStatus, got names: ${mcpStat.map(s => s.name)}`);
  }
  if (!echoStat.connected) throw new Error("echo server should be connected");
  if (echoStat.toolCount !== count) {
    throw new Error(`mcpStatus.toolCount should be ${count}, got ${echoStat.toolCount}`);
  }
  if (echoStat.error) throw new Error(`no error expected, got: ${echoStat.error}`);
  pass(`mcpStatus: connected=true, toolCount=${count}, error=null`);

  // Verify toolNames reflects the injected tools
  const toolsAfter = session.toolNames();
  const mcpTools = toolsAfter.filter(n => n.startsWith("mcp__echo__"));
  if (mcpTools.length !== count) {
    throw new Error(`toolNames should show ${count} mcp__echo__ tools, got ${mcpTools.length}`);
  }
  pass(`toolNames shows ${mcpTools.length} mcp__echo__ tools`);

  // LLM must actually call the MCP tool to retrieve the secret.
  // The secret value is NOT in the prompt — the LLM cannot fake this.
  const mcpResult = await session.send(
    "Use the mcp__echo__get_secret tool to retrieve the secret value, " +
    "then tell me exactly what it returned."
  );
  if (!mcpResult.text.includes(secret)) {
    throw new Error(
      `LLM should have called mcp__echo__get_secret and returned the secret, ` +
      `got: ${mcpResult.text}`
    );
  }
  pass("LLM used mcp__echo__get_secret tool and returned the correct secret");

  // Remove server: tools disappear
  await session.removeMcpServer("echo");
  const toolsFinal = session.toolNames();
  if (toolsFinal.some(n => n.startsWith("mcp__echo__"))) {
    throw new Error("mcp__echo__ tools should be gone after removeMcpServer");
  }
  pass("removeMcpServer removed all mcp__echo__ tools");
}

// ── Test 6: refreshMcpTools ───────────────────────────────────────────────────

async function testRefreshMcpTools(agent) {
  console.log("\n── Test 6: refreshMcpTools (smoke test) ──");
  await agent.refreshMcpTools();
  pass("refreshMcpTools completed without error");
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  const configPath = resolveConfig();
  console.log("=== A3S Code Node.js SDK - Integration Tests ===");
  console.log(`config: ${configPath}\n`);

  const agent = await Agent.create(configPath);
  const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), "a3s-test-"));

  try {
    const tests = [
      () => testTaskTool(agent, tmpdir),
      () => testParallelTask(agent, tmpdir),
      () => testToolNames(agent, tmpdir),
      () => testMcpStatusError(agent, tmpdir),
      () => testMcpInjection(agent, tmpdir),
      () => testRefreshMcpTools(agent),
    ];

    let passed = 0;
    let failed = 0;

    for (const test of tests) {
      try {
        await test();
        passed++;
      } catch (err) {
        console.error(`\n  FAILED: ${err.message}`);
        failed++;
      }
    }

    console.log("\n" + "=".repeat(48));
    if (failed === 0) {
      console.log(`=== all ${passed} tests passed ===`);
    } else {
      console.log(`=== ${passed} passed, ${failed} failed ===`);
      process.exit(1);
    }
    console.log("=".repeat(48));
  } finally {
    fs.rmSync(tmpdir, { recursive: true, force: true });
  }
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});
